mox/impl/src/lib.rs

```diff
@@ -1,245 +1,239 @@
-        Self {
-            // TODO get the span for the whole tag
-            span: name.span(),
-            name,
-            attributes,
-            children: children.into_iter().map(MoxItem::from).collect(),
+impl MoxAttr {
+    fn validate_name(name: syn_rsx::NodeName) -> syn::Result<syn::Ident> {
+        use syn::{parse::Error, punctuated::Pair, PathSegment};
+        use syn_rsx::NodeName;
+
+        let invalid_error = |span| Error::new(span, "Invalid name for an attribute");
+
+        match name {
+            NodeName::Path(syn::ExprPath {
+                attrs,
+                qself: None,
+                path: syn::Path { leading_colon: None, mut segments },
+            }) if attrs.is_empty() && segments.len() == 1 => {
+                let pair = segments.pop();
+                match pair {
+                    Some(Pair::End(PathSegment { mut ident, arguments }))
+                        if arguments.is_empty() =>
+                    {
+                        mangle_ident(&mut ident);
+                        Ok(ident)
+                    }
+                    _ => Err(invalid_error(segments.span())),
+                }
+            }
+            NodeName::Dash(punctuated) => {
+                Err(Error::new(punctuated.span(), "Dash attribute name syntax isn't supported"))
+            }
+            NodeName::Colon(punctuated) => {
+                Err(Error::new(punctuated.span(), "Colon attribute name syntax isn't supported"))
+            }
+            name => Err(invalid_error(name.span())),
         }
     }
 }
 
-struct MoxTagNoChildren {
-    span: Span,
-    name: Ident,
-    attributes: Vec<MoxAttr>,
-}
-
-impl MoxTagNoChildren {
-    fn span(&self) -> Span {
-        self.span
+fn mangle_ident(ident: &mut syn::Ident) {
+    let name = ident.to_string();
+    match name.as_str() {
+        "async" | "for" | "loop" | "type" => *ident = syn::Ident::new(&(name + "_"), ident.span()),
+        _ => (),
     }
 }
 
-impl ToTokens for MoxTagNoChildren {
-    fn to_tokens(&self, tokens: &mut TokenStream) {
-        tag_to_tokens(&self.name, &self.attributes, None, tokens);
-    }
-}
+impl TryFrom<syn_rsx::Node> for MoxExpr {
+    type Error = syn::parse::Error;
 
-impl From<SnaxSelfClosingTag> for MoxTagNoChildren {
-    fn from(SnaxSelfClosingTag { name, attributes }: SnaxSelfClosingTag) -> Self {
-        let attributes = attributes.into_iter().map(MoxAttr::from).collect();
-        Self { span: name.span(), name, attributes }
+    fn try_from(node: syn_rsx::Node) -> syn::Result<Self> {
+        use syn_rsx::NodeType::*;
+        match node.node_type {
+            Element | Attribute => Err(node_parse_error(&node, "MoxExpr")),
+            Text | Block => Ok(MoxExpr { expr: node.value.unwrap() }),
+        }
     }
 }
 
-struct MoxAttr {
-    name: Ident,
-    value: TokenStream,
+fn node_parse_error(node: &syn_rsx::Node, failed: &'static str) -> syn::parse::Error {
```

Comment:
Could `failed` be an enum we define?

Reply:
It would mimic `MoxItem` which is a code duplication for me. However i see how it can be reformed, so this isn't a problem.

Reply:
Agree it's a bit of duplication but they're not the same -- one is a parsed structure and the other is what we were trying to parse. I guess if https://doc.rust-lang.org/std/mem/struct.Discriminant.html works that would be good too, but I think it would be valuable for error reporting to have an enum that parallels MoxItem but doesn't require parsed fields.

Reply:
I believe I did most reasonable thing by defining a trait:
```rust
trait NodeConvertError {
    fn node_convert_error(node: &syn_rsx::Node) -> SynError {
        SynError::new(
            node_span(&node),
            format_args!("Cannot convert {} to {}", node.node_type, std::any::type_name::<Self>(),),
        )
    }
}

impl<T> NodeConvertError for T where T: TryFrom<syn_rsx::Node> {}
```
The usage is: `Self::node_convert_error(&node)`